File: 29_03_2024_Count_Subarrays_Where_Max_Element_Appears_at_Least_K_Times.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int countSubarrays(vector<int>& nums, int k) {
        int max_element = *max_element(nums.begin(), nums.end());
        int ans = 0;
        int start = 0;
        int max_elements_in_window = 0;
        for (int end = 0; end < nums.size(); end++) {
            if (nums[end] == max_element) {
                max_elements_in_window++;
            }
            while (max_elements_in_window == k) {
                if (nums[start] == max_element) {
                    max_elements_in_window--;
                }
                start++;
            }
            ans += start;
        }
        return ans;
    }
};

File: 29_03_2024_Count_Subarrays_Where_Max_Element_Appears_at_Least_K_Times.py
class Solution(object):
    def countSubarrays(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        max_element=max(nums)
        ans=start=max_elememts_in_window=0
        for end in range(len(nums)):
            if nums[end]==max_element:
                max_elememts_in_window+=1
            while max_elememts_in_window==k:
                if nums[start]==max_element:
                    max_elememts_in_window-=1
                start+=1
            ans+=start
        return ans

